internals/search.py

```diff
@@ -125,40 +126,46 @@ def process_queriable_field(
 # A value that a feature field can be compared against.  It can be
 # a single word or a quoted string.
 VALUE_PATTERN = r'[^" ]+|"[^"]+"'
+# Logical operators.
+# TODO(kyleju): support 'OR' logic
+LOGICAL_OPERATORS_PATTERN = r'AND|OR|NOT'
 
 # Overall, a query term can be either a structured term or a full-text term.
 # Structured terms look like: FIELD OPERATOR VALUE.
 # Full-text terms look like: SINGLE_WORD, or like: "QUOTED STRING".
 TERM_RE = re.compile(
-    '(?P<field>%s)(?P<op>%s)(?P<val>%s)\s+|(?P<textterm>%s)\s+' % (
-        FIELD_NAME_PATTERN, OPERATORS_PATTERN, VALUE_PATTERN,
-        TEXT_PATTERN),
+    '(?P<logical>%s)?\s*(?P<field>%s)(?P<op>%s)(?P<val>%s)\s+|(?P<textterm>%s)\s+' % (
+        LOGICAL_OPERATORS_PATTERN, FIELD_NAME_PATTERN, OPERATORS_PATTERN,
+        VALUE_PATTERN, TEXT_PATTERN),
     re.I)
 
 
 def process_query_term(
-    field_name: str, op_str: str, val_str: str) -> Future:
+    logical_op: str, field_name: str, op_str: str, val_str: str) -> Future:
   """Parse and run a user-supplied query, if we can handle it."""
-  query_term = field_name + op_str + val_str
-  if query_term == 'pending-approval-by:me':
-    return process_pending_approval_me_query()
-  if query_term == 'starred-by:me':
-    return process_starred_me_query()
-  if query_term == 'is:recently-reviewed':
-    return process_recent_reviews_query()
-
-  if query_term == 'owner:me':
-    return search_queries.handle_me_query_async('owner')
-  if query_term == 'editor:me':
-    return search_queries.handle_me_query_async('editor')
-  if query_term == 'can_edit:me':
-    return search_queries.handle_can_edit_me_query_async()
-  if query_term == 'cc:me':
-    return search_queries.handle_me_query_async('cc')
+
+  if logical_op != 'NOT':
+    # TODO(kyleju): support negation for the following queries.
```

Comment:
Another approach to negation would be to run the original query as-is, but then negate the result.  That could be done by getting all FeatureEntry IDs and then using a set difference operation between that and the original query in `process_query()`.   That approach should not slow down the overall query process too much so long as it is done as an async query and only done once.

Reply:
I refactored the code and will implement that for both predefined query and full-text search